Src/ILGPU/Runtime/OpenCL/CLAccelerator.cs

```diff
@@ -543,8 +544,41 @@ protected override CLKernel CreateKernel(CLCompiledKernel compiledKernel)
             new CLStream(this, ptr, responsible);
 
         /// <summary cref="Accelerator.Synchronize"/>
-        protected override void SynchronizeInternal() =>
-            DefaultStream.Synchronize();
+        protected unsafe override void SynchronizeInternal()
+        {
+            // All the events to wait on. Each event represents the completion
+            // of all operations queued prior to said event.
+            var streamEvents = new List<IntPtr>();
+            try
+            {
+                ForEachChildObject<CLStream>(stream =>
+                {
+                    // Low cost IntPtr* (cl_event*) allocation.
+                    IntPtr* resultEvent = stackalloc IntPtr[1];
+                    CLException.ThrowIfFailed(
+                        CLAPI.EnqueueBarrierWithWaitList(
+                            stream.CommandQueue,
+                            Array.Empty<IntPtr>(),
+                            resultEvent));
+
+                    // Dereference pointer so we can store it.
+                    streamEvents.Add(*resultEvent);
+                });
+
+                // Wait for all the events to fire, which would mean
+                // all operations queued on an accelerator prior to
+                // synchronization have finished.
+                if (streamEvents.Count > 0)
+                    CLException.ThrowIfFailed(CLAPI.WaitForEvents(
+                        streamEvents.ToArray()));
```

Comment:
MINOR: Our EditorConfig is saying we should add braces here.
```suggestion
                if (streamEvents.Count > 0)
                {
                    CLException.ThrowIfFailed(CLAPI.WaitForEvents(
                        streamEvents.ToArray()));
                }
```